### PR TITLE
Add compression to sfctl application upload

### PR DESCRIPTION
**Pull request by Christina-Kang**
Add functionality to allow compression on application upload. The new options are listed below:

    --compress      : Applicable only to Service Fabric application packages. Create a new
                            folder containing the compressed application package to either the
                            default location, or to the location specified by the compressed-
                            location parameter, and then upload the newly created folder.
        If there is already a compressed file generated by sfctl, it will be overwritten if this
        flag is set. An error will be returned if the directory is not an application package, or if
        it is already a compressed application package. By default, the newly created compressed
        application package will be deleted after a successful upload. If upload is not successful,
        please manually clean up the compressed package as needed. The deletion does not remove any
        empty dirs which may have been created if the compressed location parameter references non
        existent directories.

    --compressed-location : The location to put the compressed application package.
        If no location is provided, the compressed package will be placed under a newly created
        folder called sfctl_compressed_temp under the parent directory specified in the path
        argument. For example, if the path argument has value C:/FolderA/AppPkg, then the compressed
        package will be added to C:/FolderA/sfctl_compressed_temp/AppPkg.

    --keep-compressed     : Whether or not to keep the generated compressed package on successful
                                         upload completion.
        If not set, then on successful completion, the compressed app packages will be deleted. If
        upload was not successful, then the application package will always be kept in the output
        directory for re-upload.

Verified:

- [ ] Build and test CI passes
- [ ] History and readme updated to reflect changes
- [ ] Package version updated according to semantic versioning rules
- [ ] Tests modified or added, when applicable
- [ ] Updated code owners file, when applicable
- [ ] Read the [PR checklist](https://github.com/Azure/service-fabric-cli/wiki/Checklist)

### CODE REVIEW

File: src/README.rst [hunk 1]
```diff
@@ -16,6 +16,11 @@ To get started, after installation run the following:
 Change Log
 ==========
 
+Unreleased
+----------
+- Provide option to compress packages on application upload. By default, the newly generated compressed package is deleted after successful upload. (#___)
```
Comment (iyyappam) [inline src/README.rst]: >rovide  [](start = 3, length = 7)

If they are already providing compressed package, we shouldn't be returning error. It should be a no-op and return success. (I couldn't put that comment in the description, so adding it here.)

Comment (Christina-Kang) [reply src/README.rst]: If they are already providing a compressed package, then they should not use the compress option. I can make that clearer in the help text and error message. Is that what you're referring to?

File: src/sfctl/custom_app.py [hunk 2]
```diff
@@ -179,12 +183,238 @@ def upload_to_native_imagestore(sesh, endpoint, abspath, basename, #pylint: disa
...
+    """
+
+    # Check if we're dealing with a dir in _check_folder_structure_and_get_dirs instead of
+    # in this function
+
+    # Normalize slashes, etc, in app_dir and output_dir
+    app_dir = _normalize_path(app_dir)
+    output_dir = _normalize_path(output_dir)
+
+    compress_copy = IgnoreCopy()
```
Comment (iyyappam) [inline src/sfctl/custom_app.py]: If the input pkg is compressed, then we shouldn't throw error. Just copy to the output location.

File: src/sfctl/custom_app.py [hunk 3]
```diff
@@ -179,12 +183,238 @@ def upload_to_native_imagestore(sesh, endpoint, abspath, basename, #pylint: disa
...
+    :return: (str) the standardized path
+    """
+
+    path = os.path.realpath(path)  # This removes slashes at the end of the path
+    path = os.path.normpath(path)
+    path = os.path.normcase(path)
+    path = os.path.abspath(path)
+    return path
+
+def compress_package(app_dir, output_dir):
```
Comment (iyyappam) [inline src/sfctl/custom_app.py]: >compress_package [](start = 4, length = 16)

Also we should provide an option for in-place compress.

File: src/sfctl/custom_app.py [hunk 4]
```diff
@@ -179,12 +183,238 @@ def upload_to_native_imagestore(sesh, endpoint, abspath, basename, #pylint: disa
...
+
+    # A list of the service packages. This should be the absolute path
+    service_packages = []
+
+    # Parse the application manifest to find which folders should have the service manifest.
+    app_manifest_parsed = ET.parse(path_to_app_manifest).getroot()
+    for child in app_manifest_parsed:
+        # Use ends with, because the tags start with the xmlns
+        if child.tag.endswith("ServiceManifestImport"):
+            # We exect a child element that looks like:
```
Comment (jeffj6123) [inline src/sfctl/custom_app.py]: expect*

**Pull request merged**